src/client/shader.cpp

```diff
@@ -189,35 +190,28 @@ class SourceShaderCache
 
 class ShaderCallback : public video::IShaderConstantSetCallBack
 {
-	std::vector<IShaderConstantSetter*> m_setters;
+	std::vector<std::unique_ptr<IShaderConstantSetter>> m_setters;
 
 public:
-	ShaderCallback(const std::vector<IShaderConstantSetterFactory *> &factories)
+	template <typename Factories>
+	ShaderCallback(const Factories &factories)
 	{
-		for (IShaderConstantSetterFactory *factory : factories)
-			m_setters.push_back(factory->create());
-	}
-
-	~ShaderCallback()
-	{
-		for (IShaderConstantSetter *setter : m_setters)
-			delete setter;
+		for (auto &&factory : factories)
+			m_setters.push_back(std::unique_ptr<IShaderConstantSetter>(factory->create()));
 	}
 
 	virtual void OnSetConstants(video::IMaterialRendererServices *services, s32 userData) override
 	{
 		video::IVideoDriver *driver = services->getVideoDriver();
 		sanity_check(driver != NULL);
 
-		bool is_highlevel = userData;
-
-		for (IShaderConstantSetter *setter : m_setters)
-			setter->onSetConstants(services, is_highlevel);
+		for (auto &&setter : m_setters)
+			setter->onSetConstants(services);
```

Comment:
> The next part will probably be to drop all the setter-constant-factory-shader nonsense.

Reply:
What are the advantages of `&&` compared to `&`? Just wondering.

Reply:
@SmallJoker In this context it is the same, `auto &&` resolves to an **l**value reference (despite looking like an rvalue one). But would that be impossible, it would resolve to a const lvalue reference instead.